test/unittest/statistics/rtps/RTPSStatisticsTests.cpp

```diff
@@ -41,181 +47,709 @@ namespace rtps {
+    uint16_t length = 255;
+    create_writer(length, RELIABLE, TRANSIENT_LOCAL);
+
+    // writer callback through participant listener
+    ASSERT_TRUE(participant_->add_statistics_listener(participant_writer_listener, EventKind::GAP_COUNT));
+
+    // writer specific callbacks
+    ASSERT_TRUE(writer_->add_statistics_listener(writer_listener));
+
+    // add a sample to the writer history that will be delivered
+    write_small_sample(length);
+    // add a second sample and remove it to generate the gap
+    write_small_sample(length);
+    ASSERT_TRUE(writer_history_->remove_change(SequenceNumber_t{0, 2}));
+
+    // create the late joiner as VOLATILE
+    create_reader(length, RELIABLE, VOLATILE);
+
+    // match writer and reader on a dummy topic
+    match_endpoints(false, "string", "statisticsSmallTopic");
+
+    // wait for reception
+    EXPECT_TRUE(reader_->wait_for_unread_cache(Duration_t(5, 0)));
+
+    // receive the second sample
+    CacheChange_t* reader_change = nullptr;
+    ASSERT_TRUE(reader_->nextUntakenCache(&reader_change, nullptr));
+
+    // wait for acknowledgement
+    EXPECT_TRUE(writer_->wait_for_all_acked(Duration_t(1, 0)));
+    reader_->releaseCache(reader_change);
+
+    // release the listeners
+    EXPECT_TRUE(writer_->remove_statistics_listener(writer_listener));
+    EXPECT_TRUE(participant_->remove_statistics_listener(participant_writer_listener, EventKind::GAP_COUNT));
+}
+
+/*
+ * This test checks the participant discovery callbacks
+ */
+TEST_F(RTPSStatisticsTests, statistics_rpts_listener_discovery_callbacks)
+{
+    using namespace ::testing;
+    using namespace fastrtps;
+    using namespace fastrtps::rtps;
+    using namespace std;
+
+    // create the listener and set expectations
+    auto participant_listener = make_shared<MockListener>();
+    ASSERT_TRUE(participant_->add_statistics_listener(participant_listener, EventKind::DISCOVERED_ENTITY));
+
+    // check callbacks on data exchange
+    atomic_int callbacks(0);
+    ON_CALL(*participant_listener, on_entity_discovery)
+            .WillByDefault([&callbacks](const eprosima::fastdds::statistics::DiscoveryTime&)
+            {
+                ++callbacks;
+            });
+    EXPECT_CALL(*participant_listener, on_entity_discovery)
+            .Times(AtLeast(5));
+
+    // create local endpoints
+    uint16_t length = 255;
+    create_endpoints(length);
+
+    // register local endpoints
+    match_endpoints(false, "string", "statisticsSmallTopic");
+
+    // create remote endpoints and register them
+    {
+        RTPSStatisticsTestsImpl remote;
+        remote.create_participant();
+        remote.create_endpoints(length);
+        remote.match_endpoints(false, "string", "statisticsSmallTopic");
+
+        int loop = 0;
+        while ( callbacks < 5 )
+        {
+            this_thread::sleep_for(chrono::milliseconds(100));
+            if ( ++loop > 10 )
+            {
+                break;
+            }
+        }
 
```

Comment:
I am missing a check that the expected callbacks are correctly received. Also, maybe this waiting can be substituted with a block until the expected callbacks are received.

Reply:
There is a `EXPECT_CALL` above that checks if the callbacks are received.
The only purpose of the loop is to provide some time for the remote entities to be created and PDP and EDP endpoints to communicate. If we introduce a condition variable then the test will timeout if the callbacks are not performed, that is not what we want.